### PR TITLE
[PFS-239] WalkFile and ListFile return incorrect results when directories and files have the same name

### PR DESCRIPTION
**Pull request by Zhang-Muyang**

### CODE REVIEW

File: src/internal/storage/fileset/validator.go [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+var _ FileSet = &pathValidator{}
+
+type pathValidator struct {
+	fs FileSet
+}
+
+// NewPathValidator checks if there is collision between file and dir names.
+func NewPathValidator(fs FileSet) FileSet {
+	pf := &pathValidator{
```
Comment (brycemcanally) [inline src/internal/storage/fileset/validator.go:21]: nit, you could just shorten this to a return here.

File: src/internal/storage/fileset/validator.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+		fs: fs,
+	}
+	return pf
+}
+
+func (pf *pathValidator) Iterate(ctx context.Context, cb func(File) error, opts ...index.Option) error {
```
Comment (brycemcanally) [inline src/internal/storage/fileset/validator.go:27]: nit, pf -> pv

File: src/internal/storage/fileset/validator.go [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+		return errors.EnsureStack(err)
+	}
+	if retErr != nil {
+		return errors.EnsureStack(retErr)
+	}
+	return nil
+}
+
+func (pf *pathValidator) IterateDeletes(_ context.Context, _ func(File) error, opts ...index.Option) error {
+	return errors.Errorf("iterating deletes in an path validator file set")
```
Comment (brycemcanally) [inline src/internal/storage/fileset/validator.go:48]: nit, an -> a, same for shards below.

Comment (Zhang-Muyang) [reply src/internal/storage/fileset/validator.go:48]: Ah, I missed these when refactoring index filtering. I'll fix these!

File: src/internal/storage/fileset/validator.go [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+	return pf
+}
+
+func (pf *pathValidator) Iterate(ctx context.Context, cb func(File) error, opts ...index.Option) error {
+	ctx = pctx.Child(ctx, "pathIterator")
+	var prev *index.Index
+	var retErr error = nil
+	if err := pf.fs.Iterate(ctx, func(f File) error {
+		idx := f.Index()
+		if retErr == nil {
```
Comment (brycemcanally) [inline src/internal/storage/fileset/validator.go:33]: We need to change this logic a bit. We want to return as soon as an error is encountered. The original logic needs to proceed even when an error is encountered because it is also computing the size of the files. So, we should just immediately return here if `CheckIndex` returns an error.

File: src/internal/storage/fileset/validator.go [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+	"github.com/pachyderm/pachyderm/v2/src/internal/storage/fileset/index"
+)
+
+var _ FileSet = &pathValidator{}
+
+type pathValidator struct {
+	fs FileSet
+}
+
+// NewPathValidator checks if there is collision between file and dir names.
```
Comment (brycemcanally) [inline src/internal/storage/fileset/validator.go:19]: This abstraction checks for more than just file and directory path collisions, so I think we should just say `NewPathValidator validates the paths in a file set` for now.

File: src/internal/storage/fileset/validator.go (hunk 4 above)
Comment (Zhang-Muyang) [reply src/internal/storage/fileset/validator.go:33]: Got it

File: src/internal/storage/fileset/validator.go (hunk 5 above)
Comment (Zhang-Muyang) [reply src/internal/storage/fileset/validator.go:19]: Yea, I'll change that.

File: src/server/pfs/server/compaction.go [hunk 6]
```diff
@@ -218,7 +216,9 @@ func (c *compactor) Validate(ctx context.Context, taskDoer task.Doer, id fileset
 	var size int64
 	for i, result := range results {
 		if errStr == "" && i != 0 {
-			errStr = checkIndex(results[i-1].Last, result.First)
+			if indxErr := fileset.CheckIndex(results[i-1].Last, result.First); indxErr != nil {
```
Comment (brycemcanally) [inline src/server/pfs/server/compaction.go:219]: nit, I generally think it is better to just use `err` when working within a smaller scope (i.e. within the if statement), and only use specially named errors when necessary and in a larger scope.

File: src/server/pfs/server/source.go [hunk 7]
```diff
@@ -37,7 +37,7 @@ func NewSource(commitInfo *pfs.CommitInfo, fs fileset.FileSet, opts ...SourceOpt
 	}
 	s := &source{
 		commitInfo: commitInfo,
-		fileSet:    fileset.NewDirInserter(fs, sc.prefix),
+		fileSet:    fileset.NewDirInserter(fileset.NewPathValidator(fs), sc.prefix),
```
Comment (brycemcanally) [inline src/server/pfs/server/source.go:40]: I think the wrapping of the fileset with a validator should just be the first line in this function, rather than here. I think it is more maintainable that way, and it communicates intention a bit better imo (i.e. always ensure the passed in fileset goes through a validator).

File: src/server/pfs/server/testing/server_test.go [hunk 8]
```diff
@@ -7614,3 +7614,39 @@ func TestNilBranchNameStream(t *testing.T) {
...
+func TestWalkFileBug(t *testing.T) {
+	ctx := pctx.TestContext(t)
+	env := realenv.NewRealEnv(ctx, t, dockertestenv.NewTestDBConfig(t).PachConfigOption)
+
+	repo := "test"
+	require.NoError(t, env.PachClient.CreateRepo(pfs.DefaultProjectName, repo))
+
+	commit1, err := env.PachClient.StartCommit(pfs.DefaultProjectName, repo, "master")
+	require.NoError(t, env.PachClient.PutFile(commit1, "/a", &bytes.Buffer{}))
+	require.NoError(t, err)
```
Comment (brycemcanally) [inline src/server/pfs/server/testing/server_test.go:7627]: This should be after the start commit. This applies to the error checks below as well.

File: src/server/pfs/server/testing/server_test.go [hunk 9]
```diff
@@ -7614,3 +7614,39 @@ func TestNilBranchNameStream(t *testing.T) {
...
+
+	commit3, err := env.PachClient.StartCommit(pfs.DefaultProjectName, repo, "master")
+	require.NoError(t, env.PachClient.PutFile(commit3, "/a/a/a", &bytes.Buffer{}))
+	require.NoError(t, err)
+	require.NoError(t, finishCommit(env.PachClient, repo, "", commit3.Id))
+
+	commit4, err := env.PachClient.StartCommit(pfs.DefaultProjectName, repo, "master")
+	require.NoError(t, env.PachClient.PutFile(commit4, "/a/a/a/a", &bytes.Buffer{}))
+	require.NoError(t, err)
+	//require.NoError(t, finishCommit(env.PachClient, repo, "", commit4.Id))
```
Comment (brycemcanally) [inline src/server/pfs/server/testing/server_test.go:7643]: This should be removed.

File: src/server/pfs/server/testing/server_test.go [hunk 10]
```diff
@@ -7614,3 +7614,39 @@ func TestNilBranchNameStream(t *testing.T) {
...
+	info, err := env.PachClient.InspectCommit(pfs.DefaultProjectName, repo, "", commit4.Id)
+	require.NotNil(t, info)
```
Comment (brycemcanally) [inline src/server/pfs/server/testing/server_test.go:7645]: I don't think this check is necessary, just the error check is fine.

File: src/server/pfs/server/testing/server_test.go (hunk 9 above)
Comment (Zhang-Muyang) [reply src/server/pfs/server/testing/server_test.go:7643]: Yea, I'll change this. The test was more of a draft.

Comment (brycemcanally) [reply src/server/pfs/server/testing/server_test.go:7643]: Actually, why is this commented out?

File: src/server/pfs/server/testing/server_test.go [hunk 11]
```diff
@@ -7614,3 +7614,39 @@ func TestNilBranchNameStream(t *testing.T) {
 		return nil
 	}))
 }
+
+func TestWalkFileBug(t *testing.T) {
```
Comment (brycemcanally) [inline src/server/pfs/server/testing/server_test.go:7618]: A few high level things:
- This test should be changed to something like `TestWalkFileInvalid` rather than something with `Bug` in the name. `Bug` doesn't really tell someone what is being tested here.
- This test could be simplified to just creating two files (i.e. /a and /a/a) that lead to a collision rather than four.
- The test should ensure that an error is returned by inspect and walk calls that trigger collisions.

Comment (Zhang-Muyang) [reply src/server/pfs/server/testing/server_test.go:7618]: Thank you! I'll change these.

File: src/server/pfs/server/testing/server_test.go (hunk 9 above)
Comment (Zhang-Muyang) [reply src/server/pfs/server/testing/server_test.go:7643]: This was just a draft and I was testing what happens when commit is finished and left open. Since this story is about open commits so I left this commented out here to indicate the commit is not finished..

File: src/server/pfs/server/compaction.go (hunk 6 above)
Comment (Zhang-Muyang) [reply src/server/pfs/server/compaction.go:219]: I was worried `err` would cause some confusion with the outer scope err, though they don't conflict. I'll fix this if that is not a problem.

File: src/server/pfs/server/testing/server_test.go (hunk 11 above)
Comment (Zhang-Muyang) [reply src/server/pfs/server/testing/server_test.go:7618]: I changed the first two bullet points. The inspect does not actually trigger a collision in the current way. Do we want to remove inspect from the test or do we want to change the way we are validating?

File: src/internal/storage/fileset/validator.go [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+		fs: fs,
+	}
+}
+
+func (pv *pathValidator) Iterate(ctx context.Context, cb func(File) error, opts ...index.Option) error {
+	ctx = pctx.Child(ctx, "pathIterator")
+	var prev *index.Index
+	if err := pv.fs.Iterate(ctx, func(f File) error {
+		idx := f.Index()
+		if retErr := checkCollisions(prev, idx); retErr != nil {
```
Comment (brycemcanally) [inline src/internal/storage/fileset/validator.go:31]: This should be reverted back to `CheckIndex`, we want to perform the full validation. Any test that expects to be able to read an errored commit will need to be updated appropriately.

File: src/server/pfs/server/source.go [hunk 13]
```diff
@@ -31,13 +31,14 @@ type source struct {
 
 // NewSource creates a Source which emits FileInfos with the information from commit, and the entries return from fileSet.
 func NewSource(commitInfo *pfs.CommitInfo, fs fileset.FileSet, opts ...SourceOption) Source {
+	validatedFS := fileset.NewPathValidator(fs)
```
Comment (brycemcanally) [inline src/server/pfs/server/source.go:34]: nit, it is preferable to overwrite `fs` rather than define a new variable (i.e. `validatedFS`). This is more maintainable because we may add a new fileset wrapper between this and where we first use it. This is similar to why creating a subcontext usually involves overwriting the existing `ctx`.

File: src/internal/storage/fileset/validator.go [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+func (pv *pathValidator) Shards(_ context.Context, _ ...index.Option) ([]*index.PathRange, error) {
+	return nil, errors.Errorf("sharding a path validator file set")
+}
+
+func CheckIndex(prev, curr *index.Index) error {
+	if prev == nil || curr == nil {
+		return nil
+	}
+	if curr.Path == prev.Path {
+		return errors.New(fmt.Sprintf("duplicate path output by different datums (%v from %v and %v from %v)", prev.Path, prev.File.Datum, curr.Path, curr.File.Datum))
```
Comment (brycemcanally) [inline src/internal/storage/fileset/validator.go:55]: This can be shortened to just `errors.Errorf`, rather than `fmt.Sprintf` in `errors.New`.

File: src/internal/storage/fileset/validator.go [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+		fs: fs,
+	}
+}
+
+func (pv *pathValidator) Iterate(ctx context.Context, cb func(File) error, opts ...index.Option) error {
+	ctx = pctx.Child(ctx, "pathIterator")
+	var prev *index.Index
+	if err := pv.fs.Iterate(ctx, func(f File) error {
+		idx := f.Index()
+		if retErr := CheckIndex(prev, idx); retErr != nil {
```
Comment (brycemcanally) [inline src/internal/storage/fileset/validator.go:31]: nit, `retErr` -> `err`.

Comment (Zhang-Muyang) [reply src/internal/storage/fileset/validator.go:31]: Is it ok to have same name `err` in different scopes? I know they do not conflict, just want to make sure if that's wanted.

File: src/server/pfs/server/testing/server_test.go [hunk 16]
```diff
@@ -2726,15 +2726,14 @@ func TestInspectFile3(t *testing.T) {
 	require.NoError(t, err)
 	require.NoError(t, env.PachClient.PutFile(commit2, "foo", strings.NewReader(fileContent2)))
 
-	fileInfo, err = env.PachClient.InspectFile(commit2, "foo")
-	require.NoError(t, err)
-	require.NotNil(t, fileInfo)
+	_, err = env.PachClient.InspectFile(commit2, "foo")
+	// there is a collision between dir foo and file foo
```
Comment (brycemcanally) [inline src/server/pfs/server/testing/server_test.go:2730]: I would instead change this test so that there isn't a collision (i.e. change either the file or directory path).

File: src/server/pfs/server/testing/server_test.go [hunk 17]
```diff
@@ -7614,3 +7613,28 @@ func TestNilBranchNameStream(t *testing.T) {
...
+	require.NoError(t, finishCommit(env.PachClient, repo, "", commit1.Id))
+
+	commit2, err := env.PachClient.StartCommit(pfs.DefaultProjectName, repo, "master")
+	require.NoError(t, err)
+	require.NoError(t, env.PachClient.PutFile(commit2, "/a/a", &bytes.Buffer{}))
+
+	_, err = env.PachClient.InspectFile(commit2, "/")
+	require.YesError(t, err)
+
+	require.NotNil(t, env.PachClient.WalkFile(commit2, "/", func(fi *pfs.FileInfo) error {
```
Comment (brycemcanally) [inline src/server/pfs/server/testing/server_test.go:7636]: This part of the test seems like it can be removed.

Comment (Zhang-Muyang) [reply src/server/pfs/server/testing/server_test.go:7636]: Do you mean remove WalkFile?

File: src/server/pachyderm_test.go [hunk 18]
```diff
@@ -6243,6 +6185,14 @@ func TestUnionInput(t *testing.T) {
 			[]string{"bash"},
 			[]string{
 				"cp -r -L /pfs/TestUnionInput* /pfs/out",
+				"for dir in /pfs/out/*",
```
Comment (brycemcanally) [inline src/server/pachyderm_test.go:6188]: I think this could be simplified by removing the inner loop. There is only one file per directory, and the file name is not even being used in the current code.

File: src/internal/storage/fileset/validator.go (hunk 15 above)
Comment (brycemcanally) [reply src/internal/storage/fileset/validator.go:31]: Yes, that is actually desirable. Taking advantage of scopes and using the conventional names for common variables (i.e. `err` and `ctx`) is a good practice. It reduces the probability of mistakes because it narrows the scope of consideration for these common variables.

File: src/server/pfs/server/testing/server_test.go (hunk 17 above)
Comment (brycemcanally) [reply src/server/pfs/server/testing/server_test.go:7636]: Yeah, but it might just be better restructure it. I would change this to a `YesError` check and the callback function to just be a no-op (i.e. just return nil).

**Review by brycemcanally — APPROVED** [review]
lgtm
